143 dynamic poll rate

iglushchuk opened the pull request:
#143 I refactored updating data to operation and added operation for waiting for another update request. For update operation I used UIApplication background tasks, to ensure that request will be completed. 

For iOS 12+ I used NWPathMonitor, if we receive path update and It's ok for new network request, we cancel current waiting operation, update operation will fire immediately.
Update interval was changed from 12 to 6 seconds.

I actually did not add tests for cancelling operation with NWPathMonitor, because it's not easy  to directly mock NWPathMonitor, as It is final class, NWPath is struct without initializer. If we really need unit tests for this, I can add protocol and adapter for NWPathMonitor, so I can mock it and call update handlers for test cases by my self.

Looking forward for code review)

lennet commented on CriticalMass/RequestManager.swift:
```diff
@@ -120,3 +124,25 @@ public class RequestManager {
+    }
+
+    func handlePathUpdate(_ path: NWPath) {
+        if path.status == .satisfied {
```
Isn't this also get called when the network changes from Wifi to cellular?

iglushchuk replied:
I guess it does get called. When I implement adapter for NWPathMonitor, I will track only status changes

iglushchuk replied:
By the way I can refactor LocationManager into operation, not a problem

lennet commented on CriticalMass/RequestManager.swift:
```diff
@@ -7,51 +7,78 @@
 
-    init(dataStore: DataStore, locationProvider: LocationProvider, networkLayer: NetworkLayer, interval: TimeInterval = 12.0, idProvider: IDProvider, url: URL) {
+    @available(iOS 12.0, *)
+    convenience init(dataStore: DataStore, locationProvider: LocationProvider, networkLayer: NetworkLayer, interval: TimeInterval = 6.0, idProvider: IDProvider, url: URL, pathMonitor: NWPathMonitor) {
```
As you sugested, I would love to see a protocol for NWPathMonitor, this would make testing easier and also remove the second initializer

iglushchuk replied:
In progress

lennet commented on CriticalMass/RequestManager.swift:
```diff
@@ -7,51 +7,78 @@
 
     private let endpoint: URL
 
-    private var hasActiveRequest = false
```
👍 Real like that we don't need that anymore :)

iglushchuk replied:
I think the same way :)

lennet commented on CriticalMass/UpdateDataOperation.swift:
```diff
@@ -0,0 +1,57 @@
+            }
+
+            return
+        }
```
having a lot of logic in the else case and having two nested guards makes it hard to read for me. Maybe we can find a cleaner solution.

iglushchuk replied:
Agree. Moved code to separate method

lennet commented on CriticalMass/RequestManager.swift:
```diff
@@ -7,51 +7,78 @@
+        setup(pathMonitor: pathMonitor)
+    }
+
+    init(dataStore: DataStore, locationProvider: LocationProvider, networkLayer: NetworkLayer, interval: TimeInterval = 6.0, idProvider: IDProvider, url: URL) {
```
We don't wanna have a constant interval of 6 We should still use 12 if the app is in background

lennet replied:
Additionally we want to use 12 if the app is in low power mode

iglushchuk replied:
changed to 12

mltbnz commented on CriticalMass/AsyncOperation.swift:
```diff
@@ -0,0 +1,63 @@
+    
+    override var isFinished: Bool {
+        set {
+            willChangeValue(forKey: "isFinished")
```
To track state maybe an enum would be nice:
```swift
enum State: String {
    case ready, executing, finished

    fileprivate var keyPath: String {
      return "is\(rawValue.capitalized)"
    }
  }
```

mltbnz replied:
also then
```swift
var state = State.ready {
  willSet {
    willChangeValue(forKey: newValue.keyPath)
    willChangeValue(forKey: state.keyPath)
  }
  didSet {
    didChangeValue(forKey: oldValue.keyPath)
    didChangeValue(forKey: state.keyPath)
  }
}
```

iglushchuk replied:
Looks nice, I changed code to your suggestion.

mltbnz commented on CriticalMass/AsyncOperation.swift:
```diff
@@ -0,0 +1,63 @@
+
+class AsyncOperation: Operation {
+    override var isAsynchronous: Bool {
+        return true
```
You can omit the return here now 🎉

iglushchuk replied:
fixed

iglushchuk commented:
Hi! Sorry, I've been quit busy last days, so I did not have time to continue. Now everything is ok I will start coding this very soon. I mean I did not forget about it :)

iglushchuk commented:
Hi! I added tests for RequestManager and added protocol for monitoring network changes. Also I refactored LocationManager to work with operation queue the similar way I did with RequestManager. But I did not add test case for it, there is no actually any tests for location manager.

So, I can think about covering LocationManager with tests if it needs. By the way "testIDDoesntChange" for IDStoreTests fails. If you tell me what is the desired logic and behind it I can fix it.

Looking forward for code review :)

lennet commented:
Thank you! There seem to be some merge conflicts around AppController and RequestManager

iglushchuk commented:
Branch is up to date with master, no conflicts now.

lennet commented on CriticalMass/AppController.swift:
```diff
@@ -6,6 +6,7 @@
 //
 
 import UIKit
+import Network
```
I think this import statement isn't needed anymore

iglushchuk replied:
right, fixed

lennet commented on CriticalMass/AppController.swift:
```diff
@@ -38,6 +39,14 @@ class AppController {
         return NetworkOperator(networkIndicatorHelper: NetworkActivityIndicatorHelper(), dataProvider: session)
     }()
 
+    private let networkObserver: NetworkObserver? = {
```
it seems like the networkObserver isn't injected into the RequestManager

iglushchuk replied:
My fault, fixed

lennet commented on CriticalMassTests/RequestManagerTests.swift:
```diff
@@ -86,4 +87,32 @@ class RequestManagerTests: XCTestCase {
+            exp.fulfill()
+        }
+        wait(for: [exp], timeout: 4)
+    }
```
Nice!

lennet commented on CriticalMass/UpdateDataOperation.swift:
```diff
@@ -0,0 +1,62 @@
+
+        let request = PostLocationRequest()
+        networkLayer.post(request: request, bodyData: bodyData) { [weak self] result in
+            guard let self = self else { return }
```
Shouldn't we set state to finished?

iglushchuk replied:
I set state to finished in completion block after result was assigned. Do you think we should set finished before assignment?

iglushchuk replied:
If we are talking about "return" in guard, then self doesn't exists there. It means that operation queue was deallocated (with request manager) and I think we don't need to set state to finished there. By the way there is compilation error: "Variable declared in 'guard' condition is not usable in its body".

lennet replied:
oh makes sense 👍

lennet commented on CriticalMass/UpdateDataOperation.swift:
```diff
@@ -0,0 +1,62 @@
+    private func getLocationsAndChatMessages() {
+        let request = GetLocationsAndChatMessagesRequest()
+        networkLayer.get(request: request) { [weak self] result in
+            guard let self = self else { return }
```
same here

lennet commented on CriticalMass/LocationManager.swift:
```diff
@@ -24,6 +24,13 @@ class LocationManager: NSObject, CLLocationManagerDelegate, LocationProvider {
         }
     }
 
+    private let operationQueue: OperationQueue = {
```
Having a separate OperationQueue for the LocationManager is problematic because the Location might be outdated when the request is happening as both operation queues might be out of sync. We should refactor the requesting and getting a location in an Async Operation and integrate it into the RequestManagers Operation queue

iglushchuk replied:
I think I understand the problem, I need a little time to figure out how to refactor LocationManager code.

lennet commented on CriticalMassTests/RequestManagerTests.swift:
```diff
@@ -9,13 +9,14 @@
 import XCTest
 
 class RequestManagerTests: XCTestCase {
```
Can you add one more test case for recovering after regaining network access?

iglushchuk commented:
Hi! I added test for regaining network access and refactored updating location to async operation. So, looking forward for review, I'm ready for discussion)

lennet commented on CriticalMassTests/RequestManagerTests.swift:
```diff
@@ -86,4 +87,54 @@ class RequestManagerTests: XCTestCase {
+
+            setup.networkObserver.update(with: .satisfied)
+            self.wait(interval: 1) {
+                setup.networkObserver.update(with: .satisfied)
```
shouldn't be one update call be enough?

iglushchuk replied:
yeah, removed unnecessary update

lennet commented on CriticalMass/LocationProvider.swift:
```diff
@@ -17,4 +17,8 @@ public enum LocationProviderPermission: String {
+
+    func updateLocation()
+    var locationUpdateHandler: (() -> Void)? { get set }
+    var locationErrorHandler: (() -> Void)? { get set }
```
Minor: maybe we can pass a completion block into updateLocation instead of exposing the blocks publicly. We also could use Result here similar to ResultCallback

iglushchuk replied:
Nice idea, I changed updateLocation() to method with completion block

lennet commented on CriticalMass/UpdateDataOperation.swift:
```diff
@@ -0,0 +1,62 @@
+
+        let body = SendLocationPostBody(device: idProvider.id, location: currentLocation)
+        guard let bodyData = try? body.encoded() else {
+            state = .finished
```
We also need a result here

iglushchuk replied:
ok, added

stephanlindauer commented on CriticalMass/RequestManager.swift:
```diff
@@ -15,46 +15,69 @@ import os.log
     private var log = OSLog(subsystem: Bundle.main.bundleIdentifier ?? "", category: "RequestManager")
 
-    public init(dataStore: DataStore, locationProvider: LocationProvider, networkLayer: NetworkLayer, interval: TimeInterval = 12.0, idProvider: IDProvider) {
+    public init(dataStore: DataStore, locationProvider: LocationProvider, networkLayer: NetworkLayer, interval: TimeInterval = 12.0, idProvider: IDProvider, networkObserver: NetworkObserver?) {
```
so 12 remains the default?

iglushchuk replied:
I left 12 seconds as it was before refactoring, can you please explain more about 1 second interval? I mean I understand that frequent requests might be a problem for server, but how user can schedule requests every second, if we set it in code for 12 seconds? Do you thing we should increase intervals? Or there should be something like minimum waiting interval?

stephanlindauer replied:
sorry maybe i misunderstood. from the title of the pr i thought this should enable the user to dynamically set the poll interval.

iglushchuk replied:
Actually no, we don't allow users to set poll interval. The main goal sounds like: if user lost internet connection and regain it later, we retrieve location and send update request as soon as possible. But next request will be send with poll interval of 12 seconds.

stephanlindauer replied:
:+1:

lennet commented on CriticalMass/WaitOperation.swift:
```diff
@@ -0,0 +1,24 @@
+//  Copyright © 2019 Pokus Labs. All rights reserved.
+//
+
+import UIKit
```
Replacing Import UIKit with Import foundation should fix some of the SPM errors

iglushchuk replied:
fixed, check it please

lennet commented on CriticalMass/RequestManager.swift:
```diff
@@ -15,46 +15,69 @@ import os.log
+        let updateDataOperation = UpdateDataOperation(locationProvider: locationProvider,
+                                                      idProvider: idProvider,
+                                                      networkLayer: networkLayer)
+        let taskIdentifier = UIApplication.shared.beginBackgroundTask {
```
We might need to. add #if canImport(UIKit) for parts that uses UIKit Apis for  SPM compatibility

iglushchuk replied:
fixed

lennet commented on CriticalMass/RequestManager.swift:
```diff
@@ -128,3 +132,18 @@ public class RequestManager {
         }
     }
 }
+
```
The extensions also needs to be marked as @available(macOS 10.12, *) for spm compatibility

iglushchuk replied:
fixed

lennet commented:
Yeah 🙌! Thanks for your patience with the change requests and finals spm support stuff

iglushchuk commented:
Thanks for interesting issue :)
